### PR TITLE
Fixes #1275 partially

### LINKED ISSUE #1275: Generate private key and download csr
##### ISSUE TYPE
 - Feature Idea

##### COMPONENT NAME
bigip_ssl_key

##### SUMMARY
I would like to generate a private key on the F5 and download the csr to request the certificate to a CA.  
This ca be done in the GUI under System --> File Management --> SSL Certificate List , select Create. Ensure that you choose "Certificate Authority" under Issuer.

### PR DESCRIPTION
**Pull request by nitinthewiz**
There are going to be two parts to this fix. Creating this module is the first part.

### CODE REVIEW

**Comment by nitinthewiz** [comment]
@wojtek0806 I need some guidance on how to run Sanity tests locally so I don't keep running into this issue. Can we talk about this in the team meeting tomorrow? 

Also, this fixes the issue partially with a new module.

**Comment by wojtek0806** [comment]
remove `_bigip_appsvcs_extension.py` from your branch and push it

File: library/modules/bigip_ssl_csr.py [hunk 1]
```diff
@@ -0,0 +1,406 @@
...
+  - This module will create SSL CSR files on a BIG-IP. CSRs
+    require an associated SSL key to prexist on the BIG-IP.
+version_added: "f5_modules 1.2"
+options:
+  name:
+    description:
+      - The name of the csr file.
+    type: str
+    required: True
+  commonName:
```
Comment (wojtek0806) [inline library/modules/bigip_ssl_csr.py:29]: wrong parameter name, no camel case

File: library/modules/bigip_ssl_csr.py [hunk 2]
```diff
@@ -0,0 +1,406 @@
...
+                    'supported_by': 'certified'}
+
+DOCUMENTATION = r'''
+---
+module: bigip_ssl_csr
+short_description: Create SSL CSR files on the BIG-IP
+description:
+  - This module will create SSL CSR files on a BIG-IP. CSRs
+    require an associated SSL key to prexist on the BIG-IP.
+version_added: "f5_modules 1.2"
```
Comment (wojtek0806) [inline library/modules/bigip_ssl_csr.py:22]: wrong version added

File: library/modules/bigip_ssl_csr.py [hunk 3]
```diff
@@ -0,0 +1,406 @@
...
+author:
+  - Nitin Khanna (@nitinthewiz)
+'''
+
+EXAMPLES = r'''
+- name: Create an SSL csr
+  bigip_ssl_csr:
+    name: csr-name
+    key: key-name
+    commonName: csr-name
```
Comment (wojtek0806) [inline library/modules/bigip_ssl_csr.py:58]: wrong parameter

File: library/modules/bigip_ssl_csr.py [hunk 4]
```diff
@@ -0,0 +1,406 @@
...
+      user: admin
+  delegate_to: localhost
+'''
+
+RETURN = r'''
+csr_name:
+  description: The name of the CSR file that the user provided
+  returned: created
+  type: str
+  sample: csr-name
```
Comment (wojtek0806) [inline library/modules/bigip_ssl_csr.py:71]: where is the other returned parameter

File: library/modules/bigip_ssl_csr.py [hunk 5]
```diff
@@ -0,0 +1,406 @@
...
+                continue
+            else:
+                if isinstance(change, dict):
+                    changed.update(change)
+                else:
+                    changed[k] = change
+        if changed:
+            self.changes = UsableChanges(params=changed)
+            return True
+        return False
```
Comment (wojtek0806) [inline library/modules/bigip_ssl_csr.py:197]: if this module cannot update, just create and remove then this is not needed

File: library/modules/bigip_ssl_csr.py [hunk 6]
```diff
@@ -0,0 +1,406 @@
...
+        )
+        resp = self.client.api.patch(uri, json=params)
+        try:
+            response = resp.json()
+        except ValueError as ex:
+            raise F5ModuleError(str(ex))
+
+        if resp.status in [200, 201] or 'code' in response and response['code'] in [200, 201]:
+            return True
+        raise F5ModuleError(resp.content)
```
Comment (wojtek0806) [inline library/modules/bigip_ssl_csr.py:335]: again if you do not update then this is redundant

File: library/modules/bigip_ssl_csr.py [hunk 7]
```diff
@@ -0,0 +1,406 @@
...
+        )
+        resp = self.client.api.get(uri)
+        try:
+            response = resp.json()
+        except ValueError as ex:
+            raise F5ModuleError(str(ex))
+
+        if resp.status in [200, 201] or 'code' in response and response['code'] in [200, 201]:
+            return ApiParameters(params=response)
+        raise F5ModuleError(resp.content)
```
Comment (wojtek0806) [inline library/modules/bigip_ssl_csr.py:363]: we are not updating anything? if so then remove

File: library/modules/bigip_ssl_csr.py [hunk 8]
```diff
@@ -0,0 +1,406 @@
...
+        return False
+
+    def update(self):
+        self.have = self.read_current_from_device()
+        if not self.should_update():
+            return False
+        if self.module.check_mode:
+            return True
+        self.update_on_device()
+        return True
```
Comment (wojtek0806) [inline library/modules/bigip_ssl_csr.py:259]: if not updating anything remove

File: library/modules/bigip_ssl_csr.py [hunk 9]
```diff
@@ -0,0 +1,406 @@
...
+        if LooseVersion(version) < LooseVersion('14.0.0'):
+            return True
+        else:
+            return False
+
+    def present(self):
+        if self.exists():
+            return self.update()
+        else:
+            return self.create()
```
Comment (wojtek0806) [inline library/modules/bigip_ssl_csr.py:239]: if we don't update then just return False on exists() not self.update()

File: test/integration/bigip_ssl_csr.yaml [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+#
+
+- name: Metadata of bigip_ssl_csr
+  hosts: localhost
+  connection: local
+  gather_facts: false
+
+  vars:
+    __metadata__:
+      upstream: no
```
Comment (wojtek0806) [inline test/integration/bigip_ssl_csr.yaml:29]: change it to: yes

File: library/modules/bigip_ssl_csr.py [hunk 11]
```diff
@@ -0,0 +1,406 @@
...
+                    'status': ['preview'],
+                    'supported_by': 'certified'}
+
+DOCUMENTATION = r'''
+---
+module: bigip_ssl_csr
+short_description: Create SSL CSR files on the BIG-IP
+description:
+  - This module will create SSL CSR files on a BIG-IP. CSRs
+    require an associated SSL key to prexist on the BIG-IP.
```
Comment (wojtek0806) [inline library/modules/bigip_ssl_csr.py:21]: should be: preexist

File: library/modules/bigip_ssl_csr.py (hunk 1 above)
Comment (nitinthewiz) [reply library/modules/bigip_ssl_csr.py:29]: Gah! I wrote all the testing correctly, just made this documentation mistake! Fixed.

File: library/modules/bigip_ssl_csr.py (hunk 2 above)
Comment (nitinthewiz) [reply library/modules/bigip_ssl_csr.py:22]: fixed

File: library/modules/bigip_ssl_csr.py (hunk 3 above)
Comment (nitinthewiz) [reply library/modules/bigip_ssl_csr.py:58]: fixed

File: library/modules/bigip_ssl_csr.py (hunk 5 above)
Comment (nitinthewiz) [reply library/modules/bigip_ssl_csr.py:197]: Good point. Removed this code.

File: library/modules/bigip_ssl_csr.py (hunk 6 above)
Comment (nitinthewiz) [reply library/modules/bigip_ssl_csr.py:335]: Good point. Removed this code.

File: library/modules/bigip_ssl_csr.py (hunk 9 above)
Comment (nitinthewiz) [reply library/modules/bigip_ssl_csr.py:239]: yeah I set it to return False

**Comment by nitinthewiz** [comment]
> Looking at the module, and thinking more about it, we could add that key upload/download here this means we don't mess with ssl_key module, what is required to have that functionality?

The ask is not for key upload/download. It's for key creation, which should rest in ssl_key if we decide to add it in. Key upload already exists there. What would be required to add that would be simple - a key creation API endpoint exists at sys/crypto/key

**Comment by wojtek0806** [comment]
i have already clarified this with Forrest what we need. I need you to add CSR file download, look at qkview module for a solution on how to achieve that.

**Comment by nitinthewiz** [comment]
> @nitinthewiz
> i have already clarified this with Forrest what we need. I need you to add CSR file download, look at qkview module for a solution on how to achieve that.

I've added the download code based on the qkview module.

**Pull request merged**